cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -423,12 +424,47 @@ public RunRecordDetail getRun(ProgramRunId id) {
   }
 
   @Override
-  public void addApplication(ApplicationId id, ApplicationSpecification spec) {
+  public void addApplication(ApplicationId id, ApplicationSpecification spec, @Nullable String author,
+                             Long created, @Nullable String changeSummary, @Nullable String parentVersion) {
+    // Generate the creation time for the version of the app.
     TransactionRunners.run(transactionRunner, context -> {
-      getAppMetadataStore(context).writeApplication(id.getNamespace(), id.getApplication(), id.getVersion(), spec);
+      ApplicationMeta latest = getLatest(id.getNamespace(), id.getApplication());
+      if (!deployAppAllowed(parentVersion, latest)) {
+        throw new BadRequestException("Can't deploy the application because the parent version is not the " +
```

Comment:
Exceptions like this should be in the method signature as `throws BadRequestException` so that the caller knows it can happen and handle it appropriately. Right now it is getting wrapped by TransactionRunners.run() and then rethrown. There is another version of the method in TransationRunners that will rethrow the exception directly instead of wrapping it.